README.md
# jasperclient
Jasper Server REST v2 API client for publishing and running reports.

## Usage

```javascript
const jasperclient = require('jasperclient');
const fs = require('fs');

let client = new jasperclient({
    host: 'jasper.example.com',
    path: 'jasperserver',
    username: 'username',
    password: 'password',
    useBasicAuth: true,
});

// Publish a report
let publishResponse = client.reports.publish({
    path: '/Reports/Dummy',
    label: 'Dummy Report',
    datasource: '/datasources/Dummy',
    jrxml: Buffer.from(fs.readFileSync('./dummy.jrxml'),'UTF-8').toString('base64'),
    resources: [
        {
            'name': 'img.png',
            'file': {
                'fileResource': {
                    'type': 'img',
                    'label': 'img.png',
                    'content': Buffer.from(fs.readFileSync('./img.png')).toString('base64'),
                },
            },
        },
        {
            'name': 'sub.jrxml',
            'file': {
                'fileResource': {
                    'type': 'jrxml',
                    'label': 'sub.jrxml',
                    'content': Buffer.from(fs.readFileSync('./sub.jrxml'),'UTF-8').toString('base64'),
                },
            },
        }
    ],
});

// Fetch a report
let runResponse = client.reports.run({
    path: '/Reports/Dummy.pdf',
    format: 'pdf',
    params: {
        input1: 'value1',
        input2: 'value2',
    },
},{
    responseType: 'stream',
});

let writer = fs.createWriteStream('./dummy.pdf');

runResponse.data.pipe(writer);

writer.on('finish', function () {
    console.log('Successfully wrote file to ./dummy.pdf');
    // client.logout(); // only useful if useBasicAuth is false
});

writer.on('error', function (err) {
    console.log('Failed to write file: ',err);
    // client.logout(); // only useful if useBasicAuth is false
});
```

## Constructor
```javascript
@class - jasperclient -- Jasper Server REST API client for publishing and running reports.
@param { Object}  opt - Required configuration data.
@param { string}  opt.proto [http] - Protocol to use; optionally set to https.
@param { string}  opt.host - Hostname of the Jasper server.
@param {!number}  opt.port - Port number of the server (default is undefined).
@param {!string}  opt.path - Path of the server after the hostname (i.e. jasperserver).
@param {!string}  opt.username - Username to use when authenticating with the Jasper server.
@param {!string}  opt.password - Password to use with the username.
@param {!Boolean} opt.userBasicAuth - Send Basic Authorization header instead of using cookies (default is false)
```

## Methods

### login()
```javascript
@function login - Create a cookie-based session with the Jasper server.  If the username and password
                  were provided to the constructor then they can be ommitted here, and it becomes
                  unneccessary to call this method directly; it will be called automatically when
                  the first request is made.
@async
@param {!string} opt.username - Username to use when authenticating with the Jasper server.
@param {!string} opt.password - Password to use with the username.
@returns {Promise<string>} - Resolves JSESSIONID upon success.
                           - Rejects with error string on failure.
```

### logout()
```javascript
@function logout - End the current session with the Jasper server.
@async
@returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
                           - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
```

### request()
```javascript
@function request - Send a request to the jasper server.  If a username and password
                    were provided to the constructor then a login will be performed
                    if needed prior to sending this request.
@async
@param { string} method - HTTP request method to use (i.e. GET or POST).
@param { string} path - Portion of URI after hostname and path suffix.
@param {!Object} params - Optional parameters to pass with the request.
@param {!Object} data - Optional data to pass as the body of the request.
@param {!Object} opt - Additional options to apply to the Axios request (see https://github.com/axios/axios#request-config ).
@returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
                           - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
```

### resources.list()
```javascript
@function list - Fetch resource information from the server.
@param { Object} req - Details about the report you are fetching.
@param { Object} req.data - Parameters to supply with the request (see https://community.jaspersoft.com/documentation/jasperreports-server-web-services-guide/v56/searching-repository ).
@param {!Object} opt - Additional options to apply to the Axios request (see https://github.com/axios/axios#request-config ).
@returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
                           - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
```

### reports.publish()
```javascript
@function publish - Upload a report to the Jasper server.  This is limited to uploading
                    reports composed of a single .jrxml file.
@async
@param { Object} req - Details about the report you are uploading.
@param { String} req.path - Path where the report will later become available.
@param { String} req.label - Label used to identify the report on the server.
@param { String} req.datasource - URI of the datasource to use with the report.
@param { String} req.jrxml - Contents of a .jrxml file encoded as Base64.
@param {!Array<Object>} req.resources - Additional resources to be uploaded with the report. (see https://community.jaspersoft.com/documentation/tibco-jasperreports-server-rest-api-reference/v630/resource-descriptors )
@returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
                           - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
```

### reports.run()
```javascript
@function run - Fetch a report from the Jasper server.
@param { Object} req - Details about the report you are fetching.
@param { String} req.path - Path of the report on the Jasper server.
@param { String} req.format - File extension being requested.
@param {!Object} req.params - Input controls to apply to the report.
@param {!Object} opt - Additional options to apply to the Axios request (see https://github.com/axios/axios#request-config ).
@returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
                           - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
```

CHANGELOG.md
# jasperclient changes

# 2.0.0 - 2023-12-20
    - Updated dependencies
    - Began building with TypeScript (changed location of main script)

# 1.2.2 - 2023-06-16
    - Fixed login fails when connecting to Jasper Server v8.x

# 1.2.1 - 2021-01-07
    - Updated Axios to fix vulnerability

# 1.2.0 - 2020-04-09
    - Added support for basic authorization

# 1.1.1 - 2020-03-04
    - Added example use of resources to the readme
    - Fixed incorrect publish path introduced in 1.1.0

# 1.1.0 - 2020-03-04
    - Added support for multiple resources per report unit

# 1.0.1 - 2019-11-25
    - Added CHANGELOG.md
    - Fixed reports.run() when using params
    - Fixed README.md

# 1.0.0 - 2019-11-18
    - Initial Release


test/01-new.js
const expect = require('chai').expect;
const nock = require('nock');

const responses = require('./responses.js');
const jasperclient = require('../build/index');

describe('Create new jasperclient object', function () {
    
    context('without arguments or with invalid arguments', function () {
        it('should throw error', function () {
            expect( function () {
                new jasperclient();
            }).to.throw('new jasperclient() requires configuration');
        });
    });
    
});

describe('Authenticate with Jasper server using cookie based sessions', function () {
    
    beforeEach( function () {
        responses.login();
    });
    
    context('with valid credentials', function () {
        it('should obtain a session id', async function () {
            let client = new jasperclient({
                host: 'example.com',
                path: 'jasperserver',
                username: 'username',
                password: 'password',
            });
            let jsessionid = await client.login();
            expect(jsessionid).to.equal('3CF53E312EE7418CBE64363163B3CAB9');
        });
    });
    
    context('with invalid credentials', function () {
        it('should receive undefined session id', async function () {
            let client = new jasperclient({
                host: 'example.com',
                path: 'jasperserver',
                username: 'username',
                password: 'wrongpassword',
            });
            try {
                var jsessionid = await client.login();
            }
            catch (err) {
                expect(jsessionid).to.be.undefined;
            }
        });
    });
    
    
});


describe('Logout of Jasper server', function () {
    
    beforeEach( function () {
        responses.login();
        responses.logout();
    });
    
    it('should wipe cookies', async function () {
        let client = new jasperclient({
            host: 'example.com',
            path: 'jasperserver',
            username: 'username',
            password: 'password',
        });
        
        await client.login();
        
        await client.logout();
        
        expect(client.cookiejar.getCookiesSync(client.baseURL).length).to.equal(0);
    });
});


test/responses.js
(() => {
    const nock = require('nock');
    const qs = require('qs');
    const fs = require('fs');
    
    /*
    * These functions install one-time interceptors for HTTP requests.
    */
    
    module.exports = {
        
        /*
        * login - intercept a login request
        */
        login: function () {
            nock('http://example.com')
                .post('/jasperserver/j_spring_security_check')
                .reply( (uri, requestBody) => {
                    let data = qs.parse(requestBody);
                    if ( data.j_username == 'username' && data.j_password == 'password' ) {
                        return [302,'',{
                            'Location': '/jasperserver/scripts/auth/loginSuccess.json;jsessionid=3CF53E312EE7418CBE64363163B3CAB9',
                            'Set-Cookie': 'JSESSIONID=3CF53E312EE7418CBE64363163B3CAB9',
                        }];
                    }
                    else {
                        return [302,'',{
                            'Location': '/jasperserver/login.html?error=1',
                        }];
                    }
                });
        },
        
        
        /*
        * logout - intercept a logout request
        */
        logout: function () {
            nock('http://example.com')
                .get('/jasperserver/logout.html')
                .reply( (uri, requestBody) => {
                    return [200];
                });
        },
        
        
        /*
        * resourceList - intercept request for report list
        */
        resourceList: function () {
            nock('http://example.com')
                .get('/jasperserver/rest_v2/resources')
                .reply( (uri, requestBody) => {
                    let data = qs.parse(requestBody);
                    return [200,[
                        {
                            version: 20,
                            permissionMask: 2,
                            creationDate: '2019-05-31T16:32:17',
                            updateDate: '2019-07-30T14:46:04',
                            label: 'Dummy',
                            description: 'A simple dummy report',
                            uri: '/Reports/Dummy',
                            resourceType: 'reportUnit',
                        }
                    ],{}];
                });
        },
        
        /*
        * publishReport - intercept request to create or update a report
        */
        publishReport: function () {
            nock('http://example.com')
                .put('/jasperserver/rest_v2/resources/Reports/Dummy')
                .reply( (uri,requestBody) => {
                    return [
                        200,
                        {
                            version: 0,
                            permissionMask: 1,
                            creationDate: '2019-11-18T21:06:19',
                            updateDate: '2019-11-18T21:06:19',
                            label: 'Dummy Report',
                            uri: '/Reports/Dummy',
                            dataSource: { dataSourceReference: { uri: '/datasources/Dummy' } },
                            jrxml: {
                                jrxmlFileReference: { uri: '/Dummy_files/Main_jrxml' }
                            },
                            alwaysPromptControls: false,
                            controlsLayout: 'popupScreen',
                            resources: {
                                "resource": [
                                    {
                                        "name": "img.png",
                                        "file": {
                                            "fileReference": {
                                                "uri": "/Dummy_files/img.png"
                                            }
                                        }
                                    },
                                    {
                                        "name": "sub.jrxml",
                                        "file": {
                                            "fileReference": {
                                                "uri": "/Dummy_files/sub.jrxml"
                                            }
                                        }
                                    },
                                ]
                            }
                        }
                    ];
                });
        },
        
        /*
        * runReport - intercept request for a PDF report
        */
        runReport: function () {
            nock('http://example.com')
                .get('/jasperserver/rest_v2/reports/Reports/Dummy.pdf')
                .reply( (uri, requestBody) => {
                    let reader = fs.createReadStream('./test/dummy.pdf');
                    return [200,reader,{
                        'content-disposition': 'attachment; filename="Dummy.pdf"',
                        'content-type': 'application/pdf',
                        'content-length': '13264',
                    }];
                });
        },
        
        
    };
})();


test/03-reports.js
const expect = require('chai').expect;
const nock = require('nock');
const fs = require('fs');

const responses = require('./responses.js');
const jasperclient = require('../build/index');

describe('Publish a report', function () {
    
    let response;
    
    before( async function () {
        responses.login();
        responses.publishReport();
        
        let client = new jasperclient({
            host: 'example.com',
            path: 'jasperserver',
            username: 'username',
            password: 'password',
        });
        
        response = await client.reports.publish({
            path: '/Reports/Dummy',
            label: 'Dummy Report',
            datasource: '/datasources/Dummy',
            jrxml: Buffer.from(fs.readFileSync('./test/dummy.jrxml'),'UTF-8').toString('base64'),
            resources: [
                {
                    'name': 'img.png',
                    'file': {
                        'fileResource': {
                            'type': 'img',
                            'label': 'img.png',
                            'content': Buffer.from(fs.readFileSync('./test/img.png')).toString('base64'),
                        },
                    },
                },
                {
                    'name': 'sub.jrxml',
                    'file': {
                        'fileResource': {
                            'type': 'jrxml',
                            'label': 'sub.jrxml',
                            'content': Buffer.from(fs.readFileSync('./test/sub.jrxml'),'UTF-8').toString('base64'),
                        },
                    },
                }
            ],
        });
    });
    
    context('create a new report', function () {
        it('should receive report with version 0',function () {
            expect(response.data.version).to.equal(0)
        });
    });
    
    
});

describe('Run a report', function () {
    
    let response;
    
    before( async function () {
        responses.login();
        responses.runReport();
        
        let client = new jasperclient({
            host: 'example.com',
            path: 'jasperserver',
            username: 'username',
            password: 'password',
        });
    
        response = await client.reports.run({
            path: '/Reports/Dummy',
            format: 'pdf',
        },{
            responseType: 'stream',
        });
    });
    
    context('without arguments', function () {
        
        it('should have an attachment', function () {
            expect( response.headers['content-disposition'] ).to.equal('attachment; filename="Dummy.pdf"');
        });
        
        it('should have content-type of application/pdf', function () {
            expect( response.headers['content-type'] ).to.equal('application/pdf');
        });
        
        it('should receive data as a stream', function () {
            expect( typeof response.data.pipe ).to.equal('function')
        });
        
    });
    
});


src/reports.js
(() => {
    
    /*
    * @class resources - Handler for requests to /rest_v2/reports
    * @param { Object} client - An instance of jasperclient
    */
    function reports (client) {
        this.client = client;
    }
    
    /*
    * @function run - Fetch a report from the Jasper server.
    * @param { Object} req - Details about the report you are fetching.
    * @param { String} req.path - Path of the report on the Jasper server.
    * @param { String} req.format - File extension being requested.
    * @param {!Object} req.params - Input controls to apply to the report.
    * @param {!Object} opt - Additional options to apply to the Axios request (see https://github.com/axios/axios#request-config ).
    * @returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
    *                            - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
    */
    reports.prototype.run = async function (req,opt) {
        let path = '/rest_v2/reports'+req.path+'.'+req.format;
        try {
            return await this.client.request('get', path, req.params, req.data, opt );
        }
        catch (err) {
            throw(err);
        }
    };
    
    /*
    * @function publish - Upload a report to the Jasper server.  This is limited to uploading
    *                     reports composed of a single .jrxml file.
    * @async
    * @param { Object} req - Details about the report you are uploading.
    * @param { String} req.path - Path where the report will later become available.
    * @param { String} req.label - Label used to identify the report on the server.
    * @param { String} req.datasource - URI of the datasource to use with the report.
    * @param { String} req.jrxml - Contents of a .jrxml file encoded as Base64.
    * @param {!Array<Object>} req.resources - Additional resources to be uploaded with the report. (see https://community.jaspersoft.com/documentation/tibco-jasperreports-server-rest-api-reference/v630/resource-descriptors )
    * @returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
    *                            - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
    */
    reports.prototype.publish = async function (req,opt) {
        try {
            return await this.client.request('put','/rest_v2/resources'+req.path,
                {},
                Object.assign(
                    {
                        'version': req.currentVersion,
                        'overwrite': true,
                        'label': req.label,
                        'dataSource': {
                            'dataSourceReference': {
                                'uri': req.datasource,
                            },
                        },
                        'jrxml': {
                            'jrxmlFile': {
                                'type': 'jrxml',
                                'label': 'Main jrxml',
                                'content': req.jrxml,
                            },
                        },
                    },
                    !Array.isArray(req.resources) || req.resources.length == 0 ? null : {
                        'resources': {
                            'resource': req.resources,
                        },
                    },
                ),
                Object.assign({},opt,{
                    'headers': {
                        'Content-Type': 'application/repository.reportUnit+json',
                    },
                }
            ));
        }
        catch (err) {
            throw(err);
        }
    };
    
    module.exports = reports;
    
})();


src/resources.js
(() => {
    
    /*
    * @class resources - Handler for requests to /rest_v2/resources
    * @param { Object} client - An instance of jasperclient
    */
    function resources (client) {
        this.client = client;
    }
    
    /*
    * @function list - Fetch resource information from the server.
    * @async
    * @param { Object} req - Details about the report you are fetching.
    * @param { Object} req.data - Parameters to supply with the request (see https://community.jaspersoft.com/documentation/jasperreports-server-web-services-guide/v56/searching-repository ).
    * @param {!Object} opt - Additional options to apply to the Axios request (see https://github.com/axios/axios#request-config ).
    * @returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
    *                            - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
    */
    resources.prototype.list = async function (req,opt) {
        try {
            return await this.client.request('get', '/rest_v2/resources'+(typeof req.path == 'string' ? req.path : ''), req.params, req.data, opt );
        }
        catch (err) {
            throw(err);
        }
    };
    
    module.exports = resources;
    
})();


src/index.js
(() => {
    
    const axios = require('axios').default;
    const axiosCookieJarSupport = require('axios-cookiejar-support').wrapper;
    const tough = require('tough-cookie');
    const Cookie = tough.Cookie;
    const qs = require('qs');
    const Base64 = require('js-base64').Base64;
    
    /*
    
    */
    const components = {
        'resources': require('./resources.js'),
        'reports': require('./reports.js'),
    };
    
    /*
    * @class - jasperclient -- Jasper Server REST API client for publishing and running reports.
    * @param { Object}  opt - Required configuration data.
    * @param { string}  opt.proto [http] - Protocol to use; optionally set to https.
    * @param { string}  opt.host - Hostname of the Jasper server.
    * @param {!number}  opt.port - Port number of the server (default is undefined).
    * @param {!string}  opt.path - Path of the server after the hostname (i.e. jasperserver).
    * @param {!string}  opt.username - Username to use when authenticating with the Jasper server.
    * @param {!string}  opt.password - Password to use with the username.
    * @param {!Boolean} opt.userBasicAuth - Send Basic Authorization header instead of using cookies (default is false)
    */
    function jasperclient (opt) {
        
        if ( !opt ) throw new Error('new jasperclient() requires configuration');
        
        this.proto = opt.proto || 'http';
        this.host = opt.host;
        this.port = opt.port;
        this.path = opt.path;
        if ( typeof this.path != 'string' ) this.path = '';
        this.path = this.path.replace(/^\/|\/$/g,''); // remove leading/trailing slashes
        
        this.username = opt.username;
        this.password = opt.password;
        
        this.useBasicAuth = opt.useBasicAuth ? true : false;
        
        if (
            typeof this.proto != 'string'
            || typeof this.host != 'string'
        ) throw new Error('new jasperclient() requires configuration');
        
        this.baseURL = `${this.proto}://${this.host}${this.port?':'+this.port:''}/${this.path != '' ? this.path+'/' : ''}`;
        
        if ( !this.useBasicAuth ) this.cookiejar = new tough.CookieJar();
        
        this.axios = axiosCookieJarSupport(
            axios.create({
                'baseURL': this.baseURL,
                'headers': {
                    'Accept': 'application/json',
                },
                'withCredentials': true,
                'jar': this.cookiejar,
            })
        );
        
        this.resources = new components.resources(this);
        this.reports = new components.reports(this);
    }
    
    /*
    * @function _request - perform an HTTP request using Axios
    * @private
    * @async
    * @param { string} method - HTTP verb name (i.e. get, post, etc.)
    * @param { string} path - Portion of URI after hostname and path suffix.
    * @param {!Object} params - Optional parameters to pass with the request.
    * @param {!Object} data - Optional data to pass as the body of the request.
    * @param {!Object} opt - Additional options to apply to the Axios request (see https://github.com/axios/axios#request-config ).
    * @returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
    *                            - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
    */
    jasperclient.prototype._request = function (method,path,params,data,opt) {
        return new Promise( (resolve,reject) => {
            if ( this.useBasicAuth ) {
                if ( !opt ) opt = {};
                if ( !opt.headers ) opt.headers = {};
                if ( this.useBasicAuth && !opt.headers.Authorization ) opt.headers.Authorization = 'Basic '+Base64.encode( this.username+':'+this.password );
            }
            this.axios.request(Object.assign({
                method: method,
                url: path,
                params: params,
                data: data,
            },opt))
            .then( response => {
                resolve(response);
            })
            .catch( err => {
                if ( err.response && err.response.status == 302 ) {
                    resolve(err.response);
                }
                else if ( err.response ) {
                    reject(err);
                }
                else {
                    reject(err);
                }
            });
        });
    };
    
    /*
    * @function matchPath - Perform pattern matching on a path after automatically removing the base path (i.e. /jasperserver).
    * @param {   string} path - Path portion of a URI that may or may not include the base path.
    * @param {...RegExp} pattern - Used to match/capture values from the path.
    * @returns {Array|undefined|null} - Results of String.prototype.match() using the provided RegExp or undefined or null if no matches
    */
    jasperclient.prototype.matchPath = function (path,...patterns) {
        let shortPath = path;
        if ( this.path != '' ) {
            let re = new RegExp('^/'+this.path+'(.*)');
            let shortPathMatches = path.match(re);
            if ( !shortPathMatches ) return;
            shortPath = shortPathMatches[1];
        }
        for ( let pattern of patterns ) {
            let patternMatches = shortPath.match(pattern);
            if ( patternMatches ) return patternMatches;
        }
        return;
    };
    
    /*
    * @function login - Create a cookie-based session with the Jasper server.  If the username and password
    *                   were provided to the constructor then they can be ommitted here, and it becomes
    *                   unneccessary to call this method directly; it will be called automatically when
    *                   the first request is made.
    * @async
    * @param {!string} opt.username - Username to use when authenticating with the Jasper server.
    * @param {!string} opt.password - Password to use with the username.
    * @returns {Promise<string>} - Resolves JSESSIONID upon success.
    *                            - Rejects with error string on failure.
    */
    jasperclient.prototype.login = function (opt) {
        return new Promise( (resolve,reject) => {
            if ( this.useBasicAuth ) {
                reject('login() disabled when useBasicAuth is enabled');
                return;
            }
            this.cookiejar.removeAllCookiesSync();
            this._request('post','/j_spring_security_check', {}, qs.stringify({
                j_username: opt ? opt.username : this.username,
                j_password: opt ? opt.password : this.password,
            }),{
                maxRedirects: 0,
            })
            .then( async response => {
                
                if ( response.status == 302 ) {
                    let header = response.headers['set-cookie'] || [];

                    for ( let str of header ) {
                        let c = Cookie.parse(str);
                        if ( c ) await this.cookiejar.setCookie(c,this.baseURL);
                    }
                    
                    let cookie = this.cookiejar.getCookiesSync(this.baseURL).find( c => c.key == 'JSESSIONID' );

                    if (
                        cookie
                        && this.matchPath(response.headers.location,
                            '^\/scripts\/auth\/loginSuccess\.json', // v7.x
                            '^\/scripts\/visualize\/auth\/loginSuccess\.json', // v8.x
                        )
                    ) {
                        resolve(cookie.value)
                    }
                    else {
                        reject('Failed to log in');
                    }
                }
                else {
                    reject('Failed to log in');
                }
            })
            .catch(reject);
        });
    };
    
    /*
    * @function logout - End the current session with the Jasper server.
    * @async
    * @returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
    *                            - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
    */
    jasperclient.prototype.logout = async function () {
        return new Promise( (resolve,reject) => {
            if ( this.useBasicAuth ) {
                reject('logout() disabled when useBasicAuth is enabled');
                return;
            }
            this._request('get','/logout.html')
            .then( response => {
                if ( response.status == 200 ) {
                    this.cookiejar.removeAllCookiesSync();
                    resolve(response);
                }
                else {
                    reject(response);
                }
            })
            .catch(reject);
        });
    };
    
    /*
    * @function request - Send a request to the jasper server.  If a username and password
    *                     were provided to the constructor then a login will be performed
    *                     if needed prior to sending this request.
    * @async
    * @param { string} method - HTTP request method to use (i.e. GET or POST).
    * @param { string} path - Portion of URI after hostname and path suffix.
    * @param {!Object} params - Optional parameters to pass with the request.
    * @param {!Object} data - Optional data to pass as the body of the request.
    * @param {!Object} opt - Additional options to apply to the Axios request (see https://github.com/axios/axios#request-config ).
    * @returns {Promise<Object>} - Resolves an Axios response (see https://github.com/axios/axios#response-schema ).
    *                            - Rejects an Axios error (see https://github.com/axios/axios#handling-errors ).
    */
    jasperclient.prototype.request = function (method,path,params,data,opt) {
        return new Promise( async (resolve,reject) => {
            method = method.toLowerCase();
            
            if ( this.useBasicAuth ) {
                try {
                    let response = await this._request(method,path,params,data,opt);
                    resolve(response);
                    return;
                }
                catch (err) {
                    reject(err);
                    return;
                }
            }
            
            if ( this.cookiejar.getCookiesSync(this.baseURL).length > 0 ) {
                // already have cookie; attempt request
                try {
                    let response = await this._request(method,path,params,data,opt);
                    if ( response && response.status != 401 ) {
                        resolve(response);
                        return;
                    }
                }
                catch (err) {
                    reject(err);
                    return;
                }
            }
            
            // request failed; attempt login, then retry original request
            let maxAttempts = 1;
            let attempt = 1
            while ( attempt <= maxAttempts ) {
                try {
                    var jsessionid = await this.login();
                }
                catch (err) {
                    reject(err);
                    return;
                }
                
                if ( jsessionid  ) {
                    try {
                        let response = await this._request(method,path,params,data,opt);
                        resolve(response);
                        return;
                    }
                    catch (err) {
                        reject(err);
                        return;
                    }
                }
                
                attempt ++;
            }
            
            reject({ error: 'Login failed' });
            
        });
    };
    
    module.exports = jasperclient;
    
})();
